ulm-semantics/main/preprocessing/events.md

```diff
@@ -2,24 +2,101 @@
 
 module ULM-PREPROCESSING-EVENTS
     imports private COMMON-K-CELL
+    imports private RUST-ERROR-SYNTAX
     imports private RUST-PREPROCESSING-CONFIGURATION
+    imports private ULM-ENCODING-SYNTAX
     imports private ULM-PREPROCESSING-SYNTAX-PRIVATE
 
+    syntax Identifier ::= "bytes_hooks"  [token]
+                        | "data"  [token]
+                        | "empty"  [token]
+                        | "Log0"  [token]
+                        | "Log1"  [token]
+                        | "Log2"  [token]
+                        | "Log3"  [token]
+                        | "Log4"  [token]
+                        | "ulm"  [token]
+
+    syntax UkmInstruction ::= #ulmPreprocessEvent
+                                  ( method: PathInExpression
+                                  , appendLastParam: ExpressionOrError
+                                  , logIdentifier: IdentifierOrError
+                                  , eventSignature: ValueOrError
+                                  )
+
     rule 
         <k>
             ulmPreprocessEvent
                 (... fullMethodPath: Method:PathInExpression
-                , eventName: _EventName:String
+                , eventName: EventName:String
+                )
+            => #ulmPreprocessEvent
+                ( Method
+                , appendParamToBytes(data, last(Param, Params))
```

Comment:
I think you're right. I don't understand Ethereum event logging well enough to know exactly what I did here (logging was mostly ignored until yesterday (or the day before?) when it suddenly became high priority, and I didn't have time to learn enough about it). However, I think that it's doing the same thing as the SIMPLE example, and I think that it would be easy to change it to something better in the future. I'll add a TODO.